packages/astro/test/fixtures/web-component/src/components/Color.js

```diff
@@ -0,0 +1,22 @@
+export default class Color extends HTMLElement {
+	constructor() {
+		super()
+
+		this.style = document.createElement('style')
```

Comment:
where is `document` coming from? I don't see the global defined anywhere.

Reply:
The custom element is never run. This call is never evaluated. This matches the way Lit components are handled.

The entire file is safely inspected as a regular JavaScript file. Some top-level objects are shimmed so that loading the component will not throw an error.

Further more, any top-level `customElements.define` calls are intercepted, and can be used to retrieve an author-defined tag name.

Reply:
Oh ok, I'm a little confused as to what this renderer is doing then. The Lit renderer does run the component's render() and put it into a `<template shadowroot>`. It might not call the constructor, I'm not sure, that part is implemented in Lit's SSR implementation.

It would be nice to document what this renderer is doing, as I can see that it is doing some stuff, but its unclear to me what's different from this than using custom elements with no renderer. Is it just to add the script tag for you when using `client:` or more than that? Would be good to document.

Reply:
Yes. I’ve thoroughly updated the proposal @ https://github.com/snowpackjs/astro/issues/1732#user-content-proposal

There will be a separate RFC to address adding the hydrator without `client:`. However, this PR does not change the existing zero-JS expectation, and requires at least one custom element to use `client:load`.